src/server/fedavg.py

```diff
@@ -497,7 +530,57 @@ def test(self):
 
             self.test_results[self.current_epoch + 1] = results
 
+        metrics = self.evaluate()
```

Comment:
If method is a traditional FL method, a.k.a, produces a single global model, then what's the difference between:
1. Distributing the new aggregated model to testing clients and aggregating the evaluation results. Evaluation results are `corrects` and `num_samples`, which can be aggregated with weights to get the final result.
2. Evaluating the new aggregated model on the server-side.

Moreover, `self.evaluate()` is not compatible to personalized FL methods and domain generalization FL methods.

Reply:
Mathematically it is of course the same, but separating the evaluations on the clients and the server has the advantage that it allows for some more flexible usage in my opinion. Let's say we want to keep the local buffers, but would still like to see the server accuracy as if the model would be send over the network (either as a goal or as debugging measure). Only evaluating on the clients and then aggregating the weighted results would make it very hard to implement this use case without advanced trickery. Additionally, from my experience it helps having the accuracy of the server model as a debugging signal even though that it may be not the ulterior goal of the process

Reply:
Sounds fair. 

But in cases of traditional FL mehod with global buffers, FL-bench will do evaluation twice (on clients and then the server), which output the same results. That's bad for efficiency. 

I think some control codes are needed here (for identifying FL method type).

Reply:
So I solved this by allowing the option to choose how often to run the different evaluations via config options.

Reply:
Good idea!